---
title: "Welcome to PrimerMiner"
author: "Vasco Elbrecht"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Welcome to PrimerMiner}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
![](PrimerMiner_logo.png)

**PrimerMiner** is an easy batch sequence downloader, which applies clustering to barcode sequences, to have optimal reference data for metabarcoding primer development. Further it allows for _in silico_ evaluation of Primers against sequences alignments.

Extensive documentation is available on GitHub (Including [YouTube tutorials](https://www.youtube.com/watch?v=bpFMJbu-18Q&list=PLGlB7fhZ-TeUwJrZ_Dppp4bwsan7lFQPp)). In the quick guide we will highlight the usage of PrimerMiner with the sample data distributed with this package. As PrimerMiner does download data from the web wich has to be processed with additional programs, please make sure to use a directory where the data is easily accessible.

The best way to fllow this tutorial is to use the `package_tutorial.R` in the folder `Sample_Data` which comes bundled with the PrimerMiner package (latest version available on GibtHub).

[PrimerMiner wiki auf GitHub](https://github.com/VascoElbrecht/PrimerMiner/wiki)


## Downloading and batch processing of barcode sequence data

The main purpose of the PrimerMiner package is to batch download sequence data from NCBI and BOLD and then apply sequence clustering to reduce the bias generated by overrepresented sequences in the databases. Additionally the specified marker sequence can be extracted from mitochondrial genomes (if applicable).

The PrimerMiner approach has several key atvantages:

* PrimerMiner can automatically batch download and process sequences from multiple sources, for as many groups as you want.
* Target marker sequences can be automatically extracted from mitochondrial genomes.
* All sequences of each group (typically on order level) are clustered to OTUs, meaning that:
each taxon is represented by only one or in case of cryptic species a few, avoiding biases generated by over represented taxa in the database.
* PrimerMiner is taxonomy independent on Genus and species level! As sequences are clustered on similarity, it does not matter if they had wrong taxonomy assigned, as long as their Family or Order was identified correctly.
* PrimerMiner takes full advantage of available partial gene sequences and mitochondrial genomes from NCBI, BOLD and your own datasets.

To get started with downloading your first barcoding sequence data, follow the instructions in the next chapter: **[Batch downloading sequences](Batchdownload.Rmd)**


```{r, echo = T, eval=F}
setwd("")
```



